core/src/main/java/org/apache/iceberg/util/TableScanUtil.java

```diff
@@ -71,6 +78,57 @@ public static CloseableIterable<FileScanTask> splitFiles(
     return CloseableIterable.combine(splitTasks, tasks);
   }
 
+  public static <T extends FileScanTask> Iterable<CombinedScanTask> planTasks(
```

Comment:
I would consider adding `PartitionScanTask` and using it as a boundary in this method. That way, we can make this logic generic and support arbitrary scan tasks with partition info.

```
public interface PartitionScanTask extends ScanTask {
  PartitionSpec spec();
  StructLike partition();
}
```

```
public static <T extends PartitionScanTask> List<ScanTaskGroup<T>> planTaskGroups(...) {}
```

@rdblue and other reviewers, thoughts on this?

Reply:
Hmm. We also need to access file information such as length and delete files to create the weight function though.

Reply:
We can use the same mechanism as in `planTaskGroups` below.

```
Function<T, Long> weightFunc =
    task -> Math.max(task.sizeBytes(), task.filesCount() * openFileCost);
```

Reply:
In general, `planTaskGroups` is a new version of planning that is supposed to replace methods that reference and work directly with `FileScanTask`. It accepts any `ScanTask`.

Reply:
I like the `PartitionScanTask` idea, although I think we would want to consider not returning a `spec` and use a struct type instead, since the partition info may be a subset of partition fields if we can combine across them.

Reply:
@rdblue, the problem is that we actually need to know the spec. That way, we can build a projection to select required partition fields for combining. I thought `PartitionScanTask` would be implemented in non-combined tasks like `FileScanTask`. Such tasks belong to one partition only.

We may need a way to expose a combine key in `ScanTaskGroup` but I thought that's different.

What do you think?

Reply:
Oh, I see. PartitionScanTask is just a single file so we can use a partition spec. I think that's good. We may want an interface that covers both the PartitionScanTask and the combined version that exposes a struct type and tuple of partition values.

Reply:
Thanks. I've added `PartitionScanTask` for now.

Reply:
+1 for using `PartitionScanTask` for now.
Let's come up with a mechanism to propagate the key to Spark later.